jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -264,17 +347,26 @@ public Entry reserve()
             if (entries.size() >= maxEntries)
                 return null;
 
-            Entry entry = new Entry();
+            Entry entry = newEntry();
             entries.add(entry);
             return entry;
         }
     }
 
+    protected Entry newEntry()
+    {
+        if (maxMultiplex >= 0 || maxUsage >= 0)
+            return new MultiEntry();
+        return new MonoEntry();
```

Comment:
Having this as a dynamic decision does have a small cost (hopefully which the JIT can optimize out).    But it is necessary to maintain the current API.    Ultimately once we remove the deprecated methods from the base pool, then this decision can be made in a overriden method in the class that also provides the MultiEntry.    So I think this is a good compromise with the current API.

@lorban can you do some performance tests to see if the `NonoEntry` based pool is faster than the previous for things like `ByteBufferPool`.   I'm not expecting much, as it still hits an atomic int about as much as the previous atomic long... but it does avoid accessing the volatile max fields, so that could be a small win.

Reply:
I can try benchmarking `Pool` again to make sure how this change affects perf, but I already have some concerns with this `newEntry` method.

Now that there can be multiple implementations of `Entry`, all `try*` calls have become virtual. They used to be monomorphic, this implementation makes them bi-morphic as there are by default 2 classes extending from `Entry`. But anyone extending `Pool` to override this method will make those `try*` methods mega-morphic.

Since `tryAcquire` is on the fast path (it's in the hot loop of `acquire`) this change introduces an easy way to silently harm performance as there can now be a mega-morphic call on the fast path. IIRC this is the reason why the pool's strategy was written the way it was because the original proposal was using multiple implementations and was measurably slower.